### ISSUE TITLE
How to add labels to the database?

### ISSUE THREAD

**Issue by ot3** [mstenta/atmospi#21]
What is the proper way to add labels to the database so they can be added to the series?

**Comment by mstenta** [comment]
I think this would be part of #10 

See the first item in the "requirements" list on that issue:

"Create a new table for managing device labels, with the following columns..."

**Comment by ot3** [comment]
This is most likely a misunderstanding: the scope in this case is adding event-based labels, not device-based labels. I probably should have used the term "flag". Sorry for the confusion.

**Comment by mstenta** [comment]
Ah ok! I thought you meant device labels. For "annotations" there is a special "Flag" database table. Right now you have to add flags manually via a database query, though. Like this:

    INSERT INTO Flag (Device, Timestamp, Value) VALUES ('28-000005496381', 1423488425, 'This is a test annotation');

Flags reference a specific device, and have a timestamp, and a value (the annotation text).  Inputting them is manual, but once they are in the Flag table they will show up in the graphs.

We should add a bit of that info to the README.md file so it's clear about how to do that.  Eventually, I would like to add a UI for adding/editing Flags.

**Comment by mstenta** [comment]
Also, on a related note... I started working on the Device Identification task last night: #10 

I am pushing my changes to the new "devices" branch. There is a CHANGELOG.md with release notes and instructions for how to upgrade to the newer version, when it's ready, because it will require database changes and settings.py changes.  It is NOT ready yet, so I wouldn't recommend doing it... but feel free to take a look at the branch if you are interested!

**Comment by ot3** [comment]
As for capturing device info: as you may have noticed from my sample code [function getDevParams(device)], that I've used the same function to specify other device-related parameters such as line colors, opacity, legend index and chart type [e.g. spline or areaspline]. I seems to me that the presentation formats are very much device related and can be specified together with the device's alternate name.

**Comment by mstenta** [comment]
Agreed. It would be nice to have all of those as columns in the Devices database table. I am taking the first step by creating the table and refactoring the existing code to use it, with the first goal being to solve the performance issue (which you also noted in #22 )

I'm not going to combine all these other features with that set of commits because it's too much all at once, but I would be open to incorporating them separately afterwards.  It may mean you will lose those features in the interim, but you will at least have labels. :-)

Please feel free to submit patches to add the additional features after #10 is in place!

**Comment by ot3** [comment]
I have made several IMHO enhancements and/or workarounds to the presentation format. Some of these may be useful to others as well. I you are interested, I can submit an annotated screenshot.

**Comment by mstenta** [comment]
Please do! In a new issue perhaps... this one is getting long and unfocused. :-)

**Comment by ot3** [comment]
I tried the FLAG insert; works like a charm. Can you also provide the code
to delete and/or modify a FLAG?

**Comment by mstenta** [comment]
Sure...

Delete:

    DELETE FROM Flag WHERE Timestamp = 1423488425;

or, if you have more than one flag on the same timestamp, use the label:

    DELETE FROM Flag WHERE Label = 'This is a test annotation';

alternatively, use both if you want to be really specific:

    DELETE FROM Flag WHERE Timestamp = 1423488425 AND Label = 'This is a test annotation';

Modify:

    UPDATE Flag SET Label = 'This is the new annotation' WHERE Label = 'This is a test annotation';

The same WHERE clauses can be used in both to make sure you only modify/delete the one you want.  If the WHERE clause matches multiple Flags, be aware that ALL will be updated... so make sure you're specific...

These are standard SQLite queries, so you can learn more about those via Google.

**Comment by ot3** [comment]
I've added a few flags using the SQL statements provided and noticed IMHO undesirable behavior: the annotation is added to the chart legend using the device's non-human sensible ID. The desired behavior is to not add the annotations to the legend at all or if this is unavoidable to use the human-sensible name.

**Comment by mstenta** [comment]
It's unavoidable: Highcharts adds it to the legend. But, if you update to v0.2.0 you will get the human-sensible names. See https://github.com/mstenta/atmospi/releases/tag/v0.2.0

**Comment by ot3** [comment]
Thanks; I will order another Pi to create a separate testing/development environment so as not to disrupt the ongoing logging...

**Comment by mstenta** [comment]
I just tested my release notes on three Pis and they worked as expected. I was able to do it without missing a data point.

Just follow the directions in the release notes step by step, and it will go smoothly. Backup your log.db file beforehand, and commit all your custom changes to a special branch. That way if anything goes wrong you can revert Git to that branch and restore your original log.db, and you should be up and running again!

**Comment by ot3** [comment]
@mstenta " Eventually, I would like to add a UI for adding/editing Flags."

I've given this some thought and to me the most natural to enter, modify or delete flags is to click on the appropriate data point in the chart. We also need some configurable flexibility regrading a vertical offset on a flag to safeguard legibility and/or prevent overlapping flags. I have no idea if this is feasible or not.
